### PR TITLE
feat(etl): add xls etl -EUBFR-37

### PR DESCRIPTION
**Pull request by emeryro**
Add xls parsing

### CODE REVIEW

**Comment by emeryro** [comment]
Parsing of xls is working, data are correctly saved in database.
For now, other ETL (csv) haven't been updated. It will be done once this PR is validated.

File: services/ingestion/etl/budg/xls/serverless.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+             - "/*"
+
+functions:
+  parseXls:
+    handler: src/events/onParseXLS.handler
+    name: ${self:provider.stage}-${self:service}-parseXls
+    memorySize: 256
+    environment:
+      BUCKET: ${self:custom.bucketName}
+    events:
```
Comment (yhuard) [inline services/ingestion/etl/budg/xls/serverless.yml]: You could (or should) also subscribe to `${self:provider.stage}-etl-budg-xlsx` topics there ;)

File: services/ingestion/etl/budg/xls/src/events/onParseXLS.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // Extract message
+  const message = JSON.parse(snsRecord.Sns.Message);
+
+  // Check file extension
+  if (['.xls', '.xlsx'].indexOf(path.extname(message.object.key)) === -1) {
+    return callback('File extension should be .xls or .xlsx');
+  }
+
+  const s3 = new AWS.S3();
+  const XLSX = require('xlsx');
```
Comment (yhuard) [inline services/ingestion/etl/budg/xls/src/events/onParseXLS.js]: Imports/requires come first. Please also check https://drone.ne-dev.eu/ec-europa/eubfr-data-lake/206, there is a couple of linting errors out there

File: services/ingestion/etl/budg/xls/src/events/onParseXLS.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      Bucket: BUCKET,
+      Key: `${message.object.key}.ndjson`,
+      Body: dataString,
+      ContentType: 'application/x-ndjson',
+    };
+
+    s3.upload(params, err => {
+      if (err) {
+        callback(err);
+      }
```
Comment (yhuard) [inline services/ingestion/etl/budg/xls/src/events/onParseXLS.js]: Ideally, the `callback` should also be called on success with some success message

File: services/ingestion/etl/budg/xls/src/lib/transform.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    type: record['Coordinator organisation type'],
+    address: record["Coordinator's address"],
+    region: record["Coordinator's region"],
+    country: record["Coordinator's country"],
+    website: record["Coordinator's website"],
+  }];
+
+  // Preprocess partners
+  const partnerArray = [];
+  for (let i = 1; i < 39; i++) {
```
Comment (yhuard) [inline services/ingestion/etl/budg/xls/src/lib/transform.js]: Where does this `39` come from?

**Review by yhuard — DISMISSED** [review]
It's promising. Some small recommendations and one question about `39`

File: services/ingestion/etl/budg/xls/src/lib/transform.js (hunk 4 above)
Comment (emeryro) [reply services/ingestion/etl/budg/xls/src/lib/transform.js]: It simply goes from the fact that there are 38 columns for partners in their xls file... (why 38 and not something else, I don't know)

**Comment by emeryro** [comment]
Modifications have been done, ready for second review

File: services/ingestion/etl/budg/xls/src/lib/transform.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      address: record["Coordinator's address"],
+      region: record["Coordinator's region"],
+      country: record["Coordinator's country"],
+      website: record["Coordinator's website"],
+    },
+  ];
+
+  // Preprocess partners
+  const partnerArray = [];
+  for (let i = 1; i < 39; i += 1) {
```
Comment (degliwe) [inline services/ingestion/etl/budg/xls/src/lib/transform.js]: hardcoded 39.
Can't we find a creative way to count partners?

**Pull request merged**